runtime/shared_heap.c

```diff
@@ -785,6 +811,475 @@ void caml_verify_heap(caml_domain_state *domain) {
+
+    /* Now we start from evac_pool and evacuate live blocks to cur_pool */
+    while( evac_pool != NULL ) {
+      value* p = (value*)((char*)evac_pool + POOL_HEADER_SZ);
+      value* end = (value*)evac_pool + POOL_WSIZE;
+      mlsize_t wh = wsize_sizeclass[sz_class];
+      value *next = NULL;
+
+      /* Every pool in the list of evacuating pools must be evacuating */
+      CAMLassert( evac_pool->evacuating == 1 );
+
+      while (p + wh <= end) {
+        header_t hd = (header_t)atomic_load_relaxed((atomic_uintnat*)p);
+
+        /* A zero header in a shared heap pool indicates an empty space */
+        if( hd != 0 ) {
+          /* Reminder: since colours have rotated, UNMARKED indicates a MARKED
+          (i.e live) block */
+          if( Has_status_hd(hd, caml_global_heap_state.UNMARKED) ) {
+            /* live block in an evacuating pool, now we allocate it in
+                to cur_pool */
+            value* new_p = cur_pool->next_obj;
+            next = (value*)new_p[1];
+            cur_pool->next_obj = next;
+
+            if( next == NULL ) {
+              /* This pool has no more free space. Move it to unswept_full_pools
+                and then advance cur_pool onwards */
+              pool* next_pool = cur_pool->next;
+
+              heap->unswept_avail_pools[sz_class] = next_pool;
+              cur_pool->next = heap->unswept_full_pools[sz_class];
+              heap->unswept_full_pools[sz_class] = cur_pool;
+
+              cur_pool = next_pool;
+
+              CAMLassert( cur_pool->evacuating == 0 );
+            }
+
+            /* Copy the block to the new location */
+            memcpy(new_p, p, Whsize_hd(hd) * sizeof(value));
+
+            /* Set first field of p to as a forwarding pointer */
+            Field(Val_hp(p), 0) = Val_hp(new_p);
+          }
+          /* We are implicitly sweeping pools in the evacuation set and thus
+             we must remember to call finalisers for Custom blocks. */
+          else if( Has_status_hd(hd, caml_global_heap_state.GARBAGE) )
+          {
+            /* Deal with custom block finalisers */
+            if (Tag_hd (hd) == Custom_tag) {
+              void (*final_fun)(value) = Custom_ops_val(Val_hp(p))->finalize;
+              if (final_fun != NULL) final_fun(Val_hp(p));
+            }
+          }
+        } else {
+          /* In the DEBUG runtime, we should overwrite the fields of empty
+              blocks from the 2nd field onwards (the 1st field is used for
+              the free list). */
+          #ifdef DEBUG
+          for( int w = 1 ; w < wh-1 ; w++ ) {
+            Field(Val_hp(p), w) = Debug_free_major;
+          }
+          #endif
+        }
+
+        p += wh;
+      }
+
+      evac_pool = evac_pool->next;
+    }
+
+    /* A pool we can allocate in to must have empty space we can allocate
+      a new block in to else it would should have been moved to the full pools
+      list. */
+    CAMLassert(cur_pool->next_obj != NULL);
```

Comment:
What if we've filled up the last "avail" pool by evacuating? Wouldn't cur_pool be NULL then?